### PR TITLE
[SPARK-27793][SQL] Add ANSI SQL day-time and year-month interval types

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to extend Catalyst's type system by two new types that conform to the SQL standard (see SQL:2016, section 4.6.3):
- `DayTimeIntervalType` represents the day-time interval type,
- `YearMonthIntervalType` for SQL year-month interval type.

This PR only adds the two new DataType implementations, and there will be more PRs as sub-tasks of SPARK-27790 to completely support the new ANSI interval types.

### Why are the changes needed?
Spark as it is today supports an INTERVAL datatype. However this type is of very limited use. Existing interval values cannot be compared with any other interval values, or persisted to storage. Spark users request to either implement new or expand existing built-in functions which produce some sort of measures for elapsed time, such as `DATEDIFF()`. Rather than work around the edges to fill the potholes of the existing INTERVAL data type, I would like to propose to deliver a proper ANSI compliant INTERVAL type that can be introduced with minimal incompatibility, is comparable and thus sortable, and can be persisted in tables.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
1. By checking coding style via:
```
$ ./dev/scalastyle
$ ./dev/lint-java
```
2. Run the test for the default sizes:
```
$ build/sbt "test:testOnly *DataTypeSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon @srielau @bart-samwel @yaooqinn Could you take a look at this PR, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DayTimeIntervalType.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+import org.apache.spark.annotation.Unstable
+
+/**
+ * The type represents day-time intervals of the SQL standard. A day-time interval is made up
+ * of a contiguous subset of the following fields:
+ *   - SECOND, seconds within minutes and possibly fractions of a second [0..59.999999],
+ *   - MINUTE, minutes within hours [0..59],
+ *   - HOUR, hours within days [0..23],
+ *   - DAY, days in the range [0..106751991].
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DayTimeIntervalType.scala:31]: how do you get the upper limit?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/YearMonthIntervalType.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import scala.math.Ordering
+import scala.reflect.runtime.universe.typeTag
+
+import org.apache.spark.annotation.Unstable
+
+/**
+ * The type represents year-month intervals of the SQL standard. A year-month interval is made up
+ * of a contiguous subset of the following fields:
+ *   - MONTH, months within years [0..11],
+ *   - YEAR, years in the range [0..178956970].
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/YearMonthIntervalType.scala:29]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DayTimeIntervalType.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DayTimeIntervalType.scala:31]: can this be negative?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DayTimeIntervalType.scala:31]: > how do you get the upper limit?

From the formula below:
```python
>>> 2**63/((60*60*23 + 60 * 59 + 59.999999) * 1000000)
106751991.1685362
```

> can this be negative?

Interval value, yes. The field, no, according to SQL:2016:
Screenshot 2021-02-24 at 19 17 31

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/YearMonthIntervalType.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/YearMonthIntervalType.scala:29]: The same
```python
>>> 2**31/12.0
178956970.66666666
```

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon I have created the umbrellas JIRA: SPARK-27790. This PR implements the first sub-task SPARK-27793.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
Also cc @viirya @maropu @dongjoon-hyun

**Comment by cloud-fan** [comment]
@MaxGekk can you mention in the PR description that, this PR only adds the two new `DataType` implementations, and there will be more PRs to completely support the new ANSI interval types?

**Comment by MaxGekk** [comment]
Also cc @gatorsmile @rxin @marmbrus @srowen since you reviewed similar PR for TimeType: https://github.com/apache/spark/pull/25678

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @cloud-fan !

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

BTW, do we have a plan to change this in the future? I'm wondering when we can remove `Unstable`, @MaxGekk and @cloud-fan .

**Comment by yaooqinn** [comment]
LGTM.

**Review by viirya — APPROVED** [review]

**Review by maropu — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Comment by cloud-fan** [comment]
@dongjoon-hyun I think we can replace `Unstable` with `Experimental` after the first milestone is completed, which should be before 3.2.0. We can remove `Experimental` after the entire project is completed and being used by 1 or more releases.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**